source/extensions/filters/http/jwt_authn/jwks.cc

```diff
@@ -0,0 +1,254 @@
+#include "extensions/filters/http/jwt_authn/jwks.h"
+
+#include "common/common/assert.h"
+#include "common/common/base64.h"
+#include "common/json/json_loader.h"
+
+#include "extensions/filters/http/jwt_authn/utils.h"
+
+#include "openssl/bn.h"
+#include "openssl/ecdsa.h"
+#include "openssl/evp.h"
+#include "openssl/rsa.h"
+#include "openssl/sha.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+namespace {
+
+/** Class to create EVP_PKEY object from string of public key, formatted in PEM or JWKs.
+ * If it failed, status_ holds the failure reason.
+ *
+ * Usage example:
+ * EvpPkeyGetter e;
+ * bssl::UniquePtr<EVP_PKEY> pkey = e.createEvpPkeyFromStr(pem_formatted_public_key);
+ * (You can use createEvpPkeyFromJwkRSA() or createEcKeyFromJwkEC() for JWKs)
+ */
+class EvpPkeyGetter : public WithStatus {
+public:
+  // Create EVP_PKEY from PEM string
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromStr(const std::string& pkey_pem) {
+    std::string pkey_der = Base64::decode(pkey_pem);
+    if (pkey_der == "") {
+      updateStatus(Status::JwksPemBadBase64);
+      return nullptr;
+    }
+    auto rsa =
+        bssl::UniquePtr<RSA>(RSA_public_key_from_bytes(castToUChar(pkey_der), pkey_der.length()));
+    if (!rsa) {
+      updateStatus(Status::JwksPemParseError);
+    }
+    return createEvpPkeyFromRsa(rsa.get());
+  }
+
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromJwkRSA(const std::string& n, const std::string& e) {
+    return createEvpPkeyFromRsa(createRsaFromJwk(n, e).get());
+  }
+
+  bssl::UniquePtr<EC_KEY> createEcKeyFromJwkEC(const std::string& x, const std::string& y) {
+    bssl::UniquePtr<EC_KEY> ec_key(EC_KEY_new_by_curve_name(NID_X9_62_prime256v1));
+    if (!ec_key) {
+      updateStatus(Status::JwksEcCreateKeyFail);
+      return nullptr;
+    }
+    bssl::UniquePtr<BIGNUM> bn_x = createBigNumFromBase64UrlString(x);
+    bssl::UniquePtr<BIGNUM> bn_y = createBigNumFromBase64UrlString(y);
+    if (!bn_x || !bn_y) {
+      // EC public key field is missing or has parse error.
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+
+    if (EC_KEY_set_public_key_affine_coordinates(ec_key.get(), bn_x.get(), bn_y.get()) == 0) {
```

Comment:
This code is way too hairy for me to be able to review in good concious. Small mistakes here and we suddenly have slighter weaker crypto than we want. @PiotrSikora can you suggest a good review process for code like this?